plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java

```diff
@@ -468,6 +469,19 @@ public PreparedStatement buildSql(ConnectorSession session, Connection connectio
         return queryBuilder.prepareStatement(session, connection, preparedQuery);
     }
 
+    @Override
+    public PreparedQuery prepareTopNQuery(PreparedQuery preparedQuery, List<SortItem> sortItems, long limit)
+    {
+        return preparedQuery
+                .transformQuery(wrapAsSubQuery())
+                .transformQuery(tryApplyTopN(Optional.of(sortItems), OptionalLong.of(limit)));
```

Comment:
This will result in unnecessary wrapping of query as a subquery if topN cannot be applied.
Can you change flow to be more explicit here. So we return original `preparedQuery` if topN cannot be applied.

Maybe make `tryApplyTopN` take `PreparedQuery` and return `Optional<PreparedQuery>`? Or String and Optional<String>?

If you want to go with fluid approach you can go with sth like:

```
        return preparedQuery
                .transformQuery()
                .usingFunction(wrapAsSubQuery())
                .usingOptionalFunction(tryApplyTopN(Optional.of(sortItems), OptionalLong.of(limit)))
                .finish();
```

If you modify `PreparedQuery` like that:
```

    public TransformedPreparedQuery transformQuery()
    {
        return new TransformedPreparedQuery(this);
    }

    public static final class TransformedPreparedQuery
    {
        private final PreparedQuery sourceQuery;
        private Optional<PreparedQuery> currentQuery;

        public TransformedPreparedQuery(PreparedQuery query)
        {
            this.sourceQuery = query;
            this.currentQuery = Optional.of(sourceQuery);
        }

        public TransformedPreparedQuery usingFunction(Function<String, String> sqlFunction)
        {
            currentQuery = currentQuery.map(query -> query.transformQuery(sqlFunction));
            return this;
        }

        public TransformedPreparedQuery usingOptionalFunction(Function<String, Optional<String>> sqlFunction)
        {
            currentQuery = currentQuery.flatMap(
                    query -> sqlFunction
                            .apply(query.getQuery())
                            .map(result -> new PreparedQuery(result, query.getParameters())));
            return this;
        }

        public PreparedQuery finish()
        {
            return currentQuery.orElse(sourceQuery);
        }
    }

```

But I am not convinced it is worth the effort and will improve readabilty.